rules/ruleset.go

```diff
@@ -90,32 +110,149 @@ func newActiveRuleSet(
 
 // NB(xichen): could make this more efficient by keeping track of matched rules
 // at previous iteration and incrementally update match results.
-func (as *activeRuleSet) MatchAll(id []byte, from time.Time, to time.Time) MatchResult {
+func (as *activeRuleSet) MatchAll(
+	id []byte,
+	isRollupID bool,
+	from, to time.Time,
+	opts MatchOptions,
+) MatchResult {
 	var (
 		fromNanos          = from.UnixNano()
 		toNanos            = to.UnixNano()
-		currMappingResults = policy.PoliciesList{as.matchMappings(id, fromNanos)}
-		currRollupResults  = as.matchRollups(id, fromNanos)
 		nextIdx            = as.nextCutoverIdx(fromNanos)
 		nextCutoverNanos   = as.cutoverNanosAt(nextIdx)
+		currMappingResults policy.PoliciesList
+		currRollupResults  []RollupResult
 	)
-
+	if currMappingPolicies, found := as.mappingPoliciesFor(id, isRollupID, fromNanos, opts); found {
+		currMappingResults = append(currMappingResults, currMappingPolicies)
+	}
+	if opts.Mode != MappingPoliciesOnly {
+		currRollupResults = as.rollupResultsFor(id, fromNanos)
+	}
 	for nextIdx < len(as.cutoverTimesAsc) && nextCutoverNanos < toNanos {
-		nextMappingPolicies := as.matchMappings(id, nextCutoverNanos)
-		nextRollupResults := as.matchRollups(id, nextCutoverNanos)
-		currMappingResults = mergeMappingResults(currMappingResults, nextMappingPolicies)
-		currRollupResults = mergeRollupResults(currRollupResults, nextRollupResults, nextCutoverNanos)
+		if nextMappingPolicies, found := as.mappingPoliciesFor(id, isRollupID, nextCutoverNanos, opts); found {
+			currMappingResults = mergeMappingResults(currMappingResults, nextMappingPolicies)
+		}
+		if opts.Mode != MappingPoliciesOnly {
+			nextRollupResults := as.rollupResultsFor(id, nextCutoverNanos)
+			currRollupResults = mergeRollupResults(currRollupResults, nextRollupResults, nextCutoverNanos)
+		}
 		nextIdx++
 		nextCutoverNanos = as.cutoverNanosAt(nextIdx)
 	}
 
-	// The result expires when it reaches the first cutover time after t among all
+	// The result expires when it reaches the first cutover time after to among all
 	// active rules because the metric may then be matched against a different set of rules.
 	return NewMatchResult(nextCutoverNanos, currMappingResults, currRollupResults)
 }
 
-func (as *activeRuleSet) matchMappings(id []byte, timeNanos int64) policy.StagedPolicies {
-	// TODO(xichen): pool the policies.
+// mappingPoliciesFor returns the mapping policies applicable to a given id at a given
+// time. It returns true if one or more matching policies are found, and false otherwise.
+func (as *activeRuleSet) mappingPoliciesFor(
```

Comment:
IMO mapping policies are the policies dictating how the metric is going to be aggregated and retained, whereas rollup results dictate how the metrics are going to be rolled up. `matchingPoliciesFor` is not as clear as `mappingPoliciesFor` IMO since a rollup policy could also be a `matchingPolicy`.